File: src/com/droidwave/stress/Stress.java
package com.droidwave.stress;

import java.util.HashMap;
import java.util.Map;

import android.app.Activity;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.SystemClock;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.Toast;

import com.droidwave.stress.view.MirrorButton;
import com.droidwave.stress.view.StackSizeIndicator;

public class Stress extends Activity {
	private static final int DECK_SUITE_COUNT = 4;
	private static final int DECK_SIZE = 13;
	private static final int CARDS_IN_DECK = DECK_SUITE_COUNT * DECK_SIZE;
	private Player[] player = new Player[2];
	private Deck[] centerDeck = new Deck[2];
	// buttonIds[player][card]
	private int[][] buttonIds = new int[][] {
			{ R.id.Button01P1, R.id.Button02P1, R.id.Button03P1,
					R.id.Button04P1 },
			{ R.id.Button01P2, R.id.Button02P2, R.id.Button03P2,
					R.id.Button04P2 } };

	private int[] currentValue = new int[2];
	private GameMode gameMode = GameMode.SINGLEPLAYER;
	private KILevel kiLevel = KILevel.EASY;
	private static final Map<KILevel, Long> kiDelay = new HashMap<KILevel, Long>();
	static {
		kiDelay.put(KILevel.EASY, 2000l);
		kiDelay.put(KILevel.MEDIUM, 1500l);
		kiDelay.put(KILevel.HARD, 500l);
	}
	private Handler mHandler = new Handler();
	private Runnable mUpdateTimerTask = new Runnable() {
		public void run() {
			long millis = SystemClock.uptimeMillis();
			doKiStep();
			// prevent it from blocking the main thread
			int factor = 1;
			long delay = kiDelay.get(kiLevel) * factor
					- (SystemClock.uptimeMillis() - millis);
			while (delay < 0) {
				factor++;
				delay = kiDelay.get(kiLevel) * factor
						- (SystemClock.uptimeMillis() - millis);
			}
			mHandler.postDelayed(mUpdateTimerTask, delay);
		}
	};

	@Override
	public void onResume() {
		doKiStep();
		mHandler.removeCallbacks(mUpdateTimerTask);
		mHandler.postDelayed(mUpdateTimerTask, kiDelay.get(kiLevel));
		super.onResume();
	}

	private void doKiStep() {
		int playerPick;
		if (gameMode == GameMode.DEMO) {
			playerPick = (int) (Math.random() * 2);
		} else {
			playerPick = 1;
		}
		int cardPick = (int) (Math.random() * 4);
		playCard(playerPick, cardPick);
	}

	private void startKI() {
		doKiStep();
		mHandler.removeCallbacks(mUpdateTimerTask);
		mHandler.postDelayed(mUpdateTimerTask, kiDelay.get(kiLevel));
	}

	@Override
	public void onPause() {
		mHandler.removeCallbacks(mUpdateTimerTask);
		super.onPause();
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.main);
		addClickListener();
		initGame();
	}

	private void addClickListener() {
		for (int[] playerButtonIds : buttonIds) {
			for (int buttonId : playerButtonIds) {
				Button button = (Button) findViewById(buttonId);
				button.setOnClickListener(clickListener);
			}
		}
	}

	private void initGame() {
		stopKI();
		// base setup
		for (int playerNumber = 0; playerNumber < 2; playerNumber++) {
			player[playerNumber] = new Player(new Deck(DECK_SIZE,
					DECK_SUITE_COUNT));
			centerDeck[playerNumber] = new Deck(0, 0);
			int randomCard = player[playerNumber].getCardFromDeck();
			setCenterStack(playerNumber, randomCard, playerNumber);
			updateInfo(playerNumber);
		}
		player[0].setColor(Color.argb(0xff, 0xff, 0x00, 0x00));
		player[1].setColor(Color.argb(0xff, 0x00, 0x00, 0xff));
		updateInfo(0);
		updateInfo(1);
		MirrorButton playerButton = null;
		for (int p = 0; p < 2; p++) {
			int openCardId = 0;
			for (int buttonId : buttonIds[p]) {
				playerButton = (MirrorButton) findViewById(buttonId);
				playerButton.setText("" + player[p].getOpenCard(openCardId++));
				playerButton.setNotifyColor(player[p].getColor());
			}
		}
		ensurePlayability();
		if (gameMode != GameMode.MULTIPLAYER) {
			startKI();
		}
	}

	private void stopKI() {
		mHandler.removeCallbacks(mUpdateTimerTask);
	}

	private void setCenterStack(int stack, int card, int playerNumber) {
		MirrorButton stackToUpdate = null;
		if (stack == 0) {
			stackToUpdate = (MirrorButton) findViewById(R.id.ButtonStack1);
			currentValue[0] = card;
		} else {
			stackToUpdate = (MirrorButton) findViewById(R.id.ButtonStack2);
			currentValue[1] = card;
		}
		centerDeck[stack].add(card);
		stackToUpdate.setNotifyColor(player[playerNumber].getColor());
		stackToUpdate.setText("" + card);
	}

	private void playerWon() {
		if (player[0].finished() && player[1].finished()) {
			toast(R.string.game_is_a_draw);
		} else if (player[0].finished()) {
			toast((getText(R.string.player_n_wins)).toString().replaceFirst(
					"%n1%", "1"));
		} else if (player[1].finished()) {
			toast((getText(R.string.player_n_wins)).toString().replaceFirst(
					"%n1%", "2"));
		}
		initGame();
	}

	private void updateInfo(int playerNumber) {
		StackSizeIndicator info = null;
		if (playerNumber == 0) {
			info = (StackSizeIndicator) findViewById(R.id.Info1);
		} else {
			info = (StackSizeIndicator) findViewById(R.id.Info2);
		}
		int remaining = player[playerNumber].remainingCards();
		info.setRemaining((float) remaining / (CARDS_IN_DECK - 4));
	}

	private void ensurePlayability() {
		boolean playable = false;
		for (int playerNumber = 0; playerNumber < 2; playerNumber++) {
			for (int cardNumber = 0; cardNumber < 4; cardNumber++) {
				for (int currentValueNumber = 0; currentValueNumber < 2; currentValueNumber++) {
					if (Math.abs(currentValue[currentValueNumber]
							- player[playerNumber].getOpenCard(cardNumber)) % 11 == 1) {
						playable = true;
					}
				}
			}
		}
		if (!playable) {
			drawNewCenterCards();
			if (player[0].finished() || player[1].finished()) {
				playerWon();
			} else {
				ensurePlayability();
			}
		}
	}

	private void drawNewCenterCards() {
		for (int playerNumber = 0; playerNumber < 2; playerNumber++) {
			int randomCard = player[playerNumber].getCardFromDeck();
			setCenterStack(playerNumber, randomCard, playerNumber);
		}
	}

	public boolean onCreateOptionsMenu(Menu menu) {
		// Inflate the currently selected menu XML resource.
		MenuInflater inflater = getMenuInflater();
		inflater.inflate(R.layout.menu, menu);
		Log.d("menu", "menu initialized");
		return true;
	}

	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
		case R.id.new_multiplayer:
			toast(R.string.new_multiplayer_started);
			gameMode = GameMode.MULTIPLAYER;
			initGame();
			break;
		case R.id.new_singleplayer:
			toast(R.string.new_singleplayer_started);
			gameMode = GameMode.SINGLEPLAYER;
			initGame();
			break;
		case R.id.new_demo:
			toast(R.string.new_demo_started);
			gameMode = GameMode.DEMO;
			initGame();
			break;
		case R.id.easyy:
			toast(getText(R.string.ai_set_to) + " " + getText(R.string.easy));
			break;
		case R.id.medium:
			toast(getText(R.string.ai_set_to) + " " + getText(R.string.medium));
			kiLevel = KILevel.MEDIUM;
			break;
		case R.id.hard:
			toast(getText(R.string.ai_set_to) + " " + getText(R.string.hard));
			kiLevel = KILevel.HARD;
			break;
		case R.id.submenu:
		case R.id.info:
			break;
		default:
			throw new Error("Unknown menu entry clicked: " + item.getItemId());
		}
		return true;
	}

	private void toast(int resId) {
		toast(getText(resId));
	}

	private void toast(CharSequence charSequence) {
		toast("" + charSequence);
	}

	private void toast(String text) {
		Toast.makeText(this, text, Toast.LENGTH_SHORT).show();
	}

	private void playCard(int playerNumber, int cardNumber) {
		int cardValue = player[playerNumber].getOpenCard(cardNumber);
		boolean playedACard = false;
		if (Math.abs(currentValue[0] - cardValue) % 11 == 1) {
			setCenterStack(0, cardValue, playerNumber);
			playedACard = true;
		} else if (Math.abs(currentValue[1] - cardValue) % 11 == 1) {
			setCenterStack(1, cardValue, playerNumber);
			playedACard = true;
		}

		if (playedACard) {
			player[playerNumber].playOpenCard(cardNumber);
			int newCard = player[playerNumber].getOpenCard(cardNumber);
			Button button = (Button) findViewById(buttonIds[playerNumber][cardNumber]);
			button.setText("" + newCard);
			ensurePlayability();
			updateInfo(playerNumber);
			if (player[playerNumber].finished()) {
				playerWon();
			}
		}
	}

	private OnClickListener clickListener = new OnClickListener() {
		public void onClick(View v) {
			Button button = (Button) v;
			int playerNumber = getPlayerByButton(button);
			int cardNumber = getPlayedCardNumber(button, player[playerNumber]);
			playCard(playerNumber, cardNumber);
		}

		private int getPlayedCardNumber(Button button, Player player) {
			for (int p = 0; p < 2; p++) {
				for (int c = 0; c < 4; c++) {
					if (buttonIds[p][c] == button.getId()) {
						return c;
					}
				}
			}
			throw new Error("Unknown Button pressed");
		}

		private int getPlayerByButton(Button button) {
			for (int p = 0; p < 2; p++) {
				for (int c = 0; c < 4; c++) {
					if (buttonIds[p][c] == button.getId()) {
						return p;
					}
				}
			}
			throw new Error("Unknown Button pressed");
		}
	};
}

File: app/src/main/java/com/droidwave/stress/Stress.kt
package com.droidwave.stress

import java.util.Random
import java.util.Timer
import java.util.TimerTask

import android.graphics.Color
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.OnClickListener
import android.widget.Button
import android.widget.Toast

import androidx.appcompat.app.AppCompatActivity

import com.droidwave.stress.view.MirrorButton
import com.droidwave.stress.view.StackSizeIndicator

class Stress : AppCompatActivity() {
    private var timer: Timer? = null
    private val player: Array<Player> = arrayOf(Player(Deck()), Player(Deck()))
    private val centerDeck = arrayOfNulls<Deck>(2)
    // buttonIds[player][card]
    private val buttonIds = arrayOf(intArrayOf(R.id.Button01P1, R.id.Button02P1, R.id.Button03P1, R.id.Button04P1), intArrayOf(R.id.Button01P2, R.id.Button02P2, R.id.Button03P2, R.id.Button04P2))

    private val currentValue = IntArray(2)
    private var gameMode = GameMode.SINGLEPLAYER
    private var kiLevel = KILevel.EASY

    private val clickListener = object : OnClickListener {
        override fun onClick(v: View) {
            val button = v as Button
            val playerNumber = getPlayerByButton(button)
            val cardNumber = getPlayedCardNumber(button)
            playCard(playerNumber, cardNumber)
        }

        private fun getPlayedCardNumber(button: Button): Int {
            for (p in 0..1) {
                for (c in 0..3) {
                    if (buttonIds[p][c] == button.id) {
                        return c
                    }
                }
            }
            throw Error("Unknown Button pressed")
        }

        private fun getPlayerByButton(button: Button): Int {
            for (p in 0..1) {
                for (c in 0..3) {
                    if (buttonIds[p][c] == button.id) {
                        return p
                    }
                }
            }
            throw Error("Unknown Button pressed")
        }
    }

    public override fun onResume() {
        startKI()
        super.onResume()
    }

    public override fun onPause() {
        stopKI()
        super.onPause()
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)
        addClickListener()
        initGame()
    }

    private fun addClickListener() {
        for (playerButtonIds in buttonIds) {
            for (buttonId in playerButtonIds) {
                val button = findViewById<Button>(buttonId)
                button.setOnClickListener(clickListener)
            }
        }
    }

    private fun initGame() {
        stopKI()
        // base setup
        for (playerNumber in 0..1) {
            player[playerNumber] = Player(Deck())
            centerDeck[playerNumber] = Deck()
            val randomCard = player[playerNumber].cardFromDeck
            setCenterStack(playerNumber, randomCard, playerNumber)
            updateInfo(playerNumber)
        }
        player[0].color = Color.argb(0xff, 0xff, 0x00, 0x00)
        player[1].color = Color.argb(0xff, 0x00, 0x00, 0xff)
        updateInfo(0)
        updateInfo(1)
        var playerButton: MirrorButton? = null
        for (p in 0..1) {
            var openCardId = 0
            for (buttonId in buttonIds[p]) {
                playerButton = findViewById(buttonId)!!
                playerButton.text = "" + player[p].getOpenCard(openCardId++)
                playerButton.notifyColor = player[p].color
            }
        }
        ensurePlayability()
        startKI()
    }

    private fun doKiStep() {
        val playerPick: Int = if (gameMode == GameMode.DEMO) {
            Random().nextInt(2)
        } else {
            1
        }
        val cardPick = Random().nextInt(4)
        playCard(playerPick, cardPick)
    }

    private fun startKI() {
        stopKI() // clean up current timer
        if (gameMode == GameMode.MULTIPLAYER) {
            return
        }
        timer = Timer().also {
            it.scheduleAtFixedRate(object : TimerTask() {
                override fun run() {
                    this@Stress.runOnUiThread { doKiStep() }
                }
            }, kiLevel.timePerCheck, kiLevel.timePerCheck)
        }
    }

    private fun stopKI() {
        timer?.cancel()
        timer?.purge()
        timer = null
    }

    private fun setCenterStack(stack: Int, card: Int, playerNumber: Int) {
        val stackToUpdate: MirrorButton = findViewById(if (stack == 0) {
            currentValue[0] = card
            R.id.ButtonStack1
        } else {
            currentValue[1] = card
            R.id.ButtonStack2
        })
        centerDeck[stack]!!.add(card)
        stackToUpdate.notifyColor = player[playerNumber].color
        stackToUpdate.text = "" + card
    }

    private fun playerWon() {
        if (player[0].finished() && player[1].finished()) {
            toast(R.string.game_is_a_draw)
        } else if (player[0].finished()) {
            toast(getString(R.string.player_n_wins, 1))
        } else if (player[1].finished()) {
            toast(getString(R.string.player_n_wins,2))
        }
        initGame()
    }

    private fun updateInfo(playerNumber: Int) {
        val info: StackSizeIndicator = findViewById(if (playerNumber == 0) {
            R.id.Info1
        } else {
            R.id.Info2
        })
        val remaining = player[playerNumber].remainingCards()
        info.setRemaining(remaining.toFloat() / (Deck.CARDS_IN_DECK - 4))
    }

    private fun ensurePlayability() {
        for (playerNumber in 0..1) {
            for (cardNumber in 0..3) {
                for (currentValueNumber in 0..1) {
                    if (Math.abs(currentValue[currentValueNumber] - player[playerNumber].getOpenCard(cardNumber)) % 11 == 1) {
                        return
                    }
                }
            }
        }
        drawNewCenterCards()
        if (player[0].finished() || player[1].finished()) {
            playerWon()
        } else {
            ensurePlayability()
        }
    }

    private fun drawNewCenterCards() {
        for (playerNumber in 0..1) {
            val randomCard = player[playerNumber].cardFromDeck
            setCenterStack(playerNumber, randomCard, playerNumber)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.new_multiplayer -> {
                toast(R.string.new_multiplayer_started)
                gameMode = GameMode.MULTIPLAYER
                initGame()
            }
            R.id.new_singleplayer -> {
                toast(R.string.new_singleplayer_started)
                gameMode = GameMode.SINGLEPLAYER
                initGame()
            }
            R.id.new_demo -> {
                toast(R.string.new_demo_started)
                gameMode = GameMode.DEMO
                initGame()
            }
            R.id.easyy -> {
                kiLevel = KILevel.EASY
                toast(getString(R.string.ai_set_to, getString(kiLevel.nameResId)))
            }
            R.id.medium -> {
                kiLevel = KILevel.MEDIUM
                toast(getString(R.string.ai_set_to, getString(kiLevel.nameResId)))
            }
            R.id.hard -> {
                kiLevel = KILevel.HARD
                toast(getString(R.string.ai_set_to, getString(kiLevel.nameResId)))
            }
            R.id.submenu, R.id.info -> {
            }
            else -> throw Error("Unknown menu entry clicked: " + item.itemId)
        }
        return true
    }

    private fun toast(resId: Int) {
        toast(getText(resId))
    }

    private fun toast(charSequence: CharSequence) {
        toast("" + charSequence)
    }

    private fun toast(text: String) {
        Toast.makeText(this, text, Toast.LENGTH_SHORT).show()
    }

    private fun playCard(playerNumber: Int, cardNumber: Int) {
        val cardValue = player[playerNumber].getOpenCard(cardNumber)
        var playedACard = false
        if (Math.abs(currentValue[0] - cardValue) % 11 == 1) {
            setCenterStack(0, cardValue, playerNumber)
            playedACard = true
        } else if (Math.abs(currentValue[1] - cardValue) % 11 == 1) {
            setCenterStack(1, cardValue, playerNumber)
            playedACard = true
        }

        if (playedACard) {
            player[playerNumber].playOpenCard(cardNumber)
            val newCard = player[playerNumber].getOpenCard(cardNumber)
            val button = findViewById<Button>(buttonIds[playerNumber][cardNumber])
            button.text = "" + newCard
            ensurePlayability()
            updateInfo(playerNumber)
            if (player[playerNumber].finished()) {
                playerWon()
            }
        }
    }
}